crypto/lms/lms_verify.c

```diff
@@ -0,0 +1,168 @@
+ *
+ * @param paths An array of bytes representing the hash values associated with
+ *              the path through the tree from the leaf associated with the
+ *              LM-OTS signature to the root public key node.
+ * @param n The hash output size (The size of each path in |paths|)
+ * @param nodenum The leaf index node number. The root node had a value of 1
+ *                Each subsequent level has nodes in the range 2^h...2^(h+1)-1
+ * @param ctx A EVP_MD_CTX object used for calculations
+ * @param ctxI A EVP_MD_CTX object containing an unfinalised H(I)
+ * @param Tc Contains H(I || u32str(node_num) || u16str(D_LEAF) || Kc) on input,
+ *           and on output returns the calculated candidate public key.
+ * @returns 1 on success, or 0 otherwise.
+ */
+static
+int lms_sig_compute_tc_from_path(const unsigned char *paths, uint32_t n,
+                                 uint32_t node_num,
+                                 EVP_MD_CTX *ctx, EVP_MD_CTX *ctxI,
+                                 unsigned char *Tc)
+{
+    int ret = 0;
+    unsigned char qbuf[4];
+    const unsigned char *path = paths;
+
+    /*
+     * Calculate the public key Tc using the path
+     * The root hash is the hash of its 2 childrens Hash values.
+     * A child hash for each level is passed in by paths, and we have
+     * a leaf value that can be used with the path to calculate the parent
+     * hash.
+     */
+    while (node_num > 1) {
+        /* At each level the path contains either the left or right child */
+        int odd = node_num & 1;
+
+        node_num = node_num >> 1; /* get the parent node_num */
+        U32STR(qbuf, node_num);
+
+        /*
+         * Calculate Tc as either
+         *   Tc(parent) = H(I || node_q || 0x8383 || paths[i][n] || Tc(right) OR
+         *   Tc(parent) = H(I || node_q || 0x8383 || Tc(left) || paths[i][n])
+         */
+        if (!EVP_MD_CTX_copy_ex(ctx, ctxI)
+                || !EVP_DigestUpdate(ctx, qbuf, sizeof(qbuf))
+                || !EVP_DigestUpdate(ctx,
+                                     &OSSL_LMS_D_INTR, sizeof(OSSL_LMS_D_INTR)))
+            goto err;
+
+        if (odd) {
+            if (!EVP_DigestUpdate(ctx, path, n)
+                || !EVP_DigestUpdate(ctx, Tc, n))
+                goto err;
+        } else {
+            if (!EVP_DigestUpdate(ctx, Tc, n)
+                || !EVP_DigestUpdate(ctx, path, n))
+                goto err;
+        }
+        /*
+         * Tc = parent Hash, which is either the left or right child for the next
+         * level up (node_num determines if it is left or right).
+         */
+        if (!EVP_DigestFinal_ex(ctx, Tc, NULL))
+            goto err;
+        path += n;
+    }
+    ret = 1;
+err:
+    return ret;
+}
+
+/*
+ * @brief LMS signature verification.
+ * See RFC 8554 Section 5.4.2. Algorithm 6: Steps 3 & 4
+ *
+ * The passed in |lms_sig| and |pub| need to exist until
+ * ossl_lms_sig_verify_final() is called, since the final may be delayed until
```

Comment:
The function referenced here does not exist

Reply:
Nice catch, this is part of the HSS stuff I think.